카테고리별 가게 목록 조회

TIP-Santa opened the pull request:
#### 시나리오
---
카테고리 아이디를 획득
해당 카테고리 에 포함된 가게 목록을 출력
- 가게 이름
- 최소 주문 금액
- 목록은 페이징 처리되어 출력 (RequestParam)
  - page
  - limit

#### 테스트
---
성공 테스트
- 가게 목록 반환
- 빈 목록 반환
실패 테스트
- 선택된 카테고리가 존재하지 않는 경우

TIP-Santa committed:
#64 feat 카테고리 별 가게 목록 조회
- GatStoreRespDto : response 필드 추가
- Store : Builder 에 id 추가
- StoreCategoryRepository : findByCategoryId 추가
- StoreController : 카테고리 별 가게 목록 조회 기능 추가
- StoreService : 카테고리 별 가게 목록 조회 기능 추가

TIP-Santa committed:
#63 feat 카테고리 별 가게 목록 조회
- GatStoreRespDto : response 필드 추가
- Store : Builder 에 id 추가
- StoreCategoryRepository : findByCategoryId 추가
- StoreController : 카테고리 별 가게 목록 조회 기능 추가
- StoreService : 카테고리 별 가게 목록 조회 기능 추가

TIP-Santa committed:
#64 test(StoreServiceTest)
성공 테스트
- 가게 조회 성공 테스트
  - 가게 목록 정상 반환
  - 빈 목록 반환

TIP-Santa committed:
#63 test(StoreServiceTest)
성공 테스트
- 가게 조회 성공 테스트
  - 가게 목록 정상 반환
  - 빈 목록 반환

TIP-Santa committed:
#64 feat(StoreController, StoreService)
- StoreController : Page 기능 추가
- StoreService : Page 기능 추가
refactor(StoreCategoryRepository)
- 반환 타입 List > Page 변경

TIP-Santa committed:
pull > merge

TIP-Santa committed:
Issue#64 refactor(StoreService) : 유효성 검사를 우선으로 실시하여 불필요한 로직 최소화

TIP-Santa committed:
Issue#64 test(StoreServiceTest) : 실패 테스트 - 존재하지 않는 카테고리 추가

TIP-Santa committed:
dev pull > merge

JIMINJI1 reviewed (approved):
확인했습니다.

TIP-Santa committed:
refactor (StoreController)

빠진 / 추가

yaewon0411 commented on src/main/java/com/my/foody/domain/store/service/StoreService.java:
```diff
@@ -86,4 +89,14 @@ public Store findActivateStoreByIdOrFail(Long storeId){
+        }
+        Pageable pageable = PageRequest.of(page, limit);
+        Page<StoreCategory> storeCategories = storeCategoryRepository.findByCategoryId(categoryId, pageable);
+        return storeCategories.map(storeCategory -> new GetStoreRespDto(storeCategory.getStore()));
```
여기서 n+1이 발생합니다
페이징이 들어가기 때문에 프로젝션을 사용할 수 있습니다
현재는 yaml에서 전역적으로 배치 크기를 설정했기 때문에 크게 문제는 없겠다만 성능상의 이슈가 발생할 수 있음을 말씀드립니다

TIP-Santa replied:
감사합니다.
프로젝션 사용하여 수정해보았습니다.

yaewon0411 commented on src/main/java/com/my/foody/domain/store/dto/resp/GetStoreRespDto.java:
```diff
@@ -8,10 +8,13 @@
 @NoArgsConstructor
 public class GetStoreRespDto {
     private String name;
+    private Long minOrderAmount;
+
     private boolean isDeleted;
 
     public GetStoreRespDto(Store store) {
         this.name = store.getName();
         this.isDeleted = store.getIsDeleted();
+        this.minOrderAmount = store.getMinOrderAmount();
     }
 }
```
클라이언트에 응답 반환할 때 Page<GetStoreRespDto> 이렇게 반환하고 있습니다

문제는 없지만 너무 많은 page 메타 데이터가 나가게 됩니다
그래서 아래와 같이 필요한 필드만 뽑아서 내보내는 방식을 제안합니다

```
@NoArgsConstructor
@Getter
public class OrderListRespDto {
    public OrderListRespDto(Page<OrderProjectionRespDto> orderList) {
        this.orderList = orderList.stream().map(OrderRespDto::new).toList();
        this.pageInfo = new PageInfo(orderList);
    }

    private List<OrderRespDto> orderList;
    private PageInfo pageInfo;

    @NoArgsConstructor
    @Getter
    @AllArgsConstructor
    public static class PageInfo {
        public PageInfo(Page<OrderProjectionRespDto> page) {
            this.pageNumber = page.getNumber();
            this.pageSize = page.getSize();
            this.totalElements = page.getTotalElements();
            this.totalPages = page.getTotalPages();
            this.isFirst = page.isFirst();
            this.isLast = page.isLast();
            this.hasNext = page.hasNext();
            this.hasPrevious = page.hasPrevious();
        }

        private int pageNumber;        // 현재 페이지 번호
        private int pageSize;          // 페이지당 항목 수
        private long totalElements;    // 전체 항목 수
        private int totalPages;        // 전체 페이지 수
        private boolean isFirst;       // 첫 페이지 여부
        private boolean isLast;        // 마지막 페이지 여부
        private boolean hasNext;       // 다음 페이지 존재 여부
        private boolean hasPrevious;   // 이전 페이지 존재 여부
    }
```

TIP-Santa replied:
알려주셔서 감사합니다.
StoreListRespDto 만들어서 사용하였습니다.

yaewon0411 commented on src/test/java/com/my/foody/domain/store/service/StoreServiceTest.java:
```diff
@@ -193,4 +194,95 @@ public void testGetAllStore_Success() {
+        verify(storeCategoryRepository, times(1)).findByCategoryId(categoryId, pageable);
+    }
+
+    @DisplayName("가게 조회 실패 테스트 - 존재하지 않는 카테고리")
+    @Test
+    public void testGetStoreByCategory_Fail_NotFoundCategory() {
+        Long categoryId = 999L; // 존재하지 않는 카테고리 ID
```
이 테스트 케이스가 실패했습니다
        when(storeRepository.countByOwnerId(ownerId)).thenReturn(2L);

여기서 2L이 아니라 ownerId를 넣어주시면 될 듯합니다!!

TIP-Santa replied:
해당 코드에는 권한이 필요없어서 owner관련 코드가 들어있지 않습니다. ㅠㅠ

yaewon0411 reviewed (commented):
최신 dev 반영해주세요!

TIP-Santa committed:
dev full > merge

TIP-Santa committed:
chore(MenuService) : 오타 수정

TIP-Santa committed:
#64 feat(StoreCategoryProjection) : 프로젝션 기능 추가

TIP-Santa committed:
#64 refactor 리뷰 반영
- StoreService
 : 프로젝션 사용 관련 코드 수정
- GetStoreRespDto
 : 프로젝션 사용 관련 GetStoreRespDto 추가
- StoreCategoryRepository
 : 프로젝션 사용 관련 쿼리 추가

TIP-Santa committed:
#64 test 리뷰 반영
- StoreServiceTest
 : 프로젝션 사용 관련 테스트 코드 일부 수정

TIP-Santa committed:
chore full > merge

TIP-Santa committed:
#64 feat 리뷰 반영
- StoreListRespDto
 : StoreListRespDto 기능 추가

TIP-Santa committed:
#64 refactor 리뷰 반영
- StoreController
 : StoreListRespDto 사용 관련 코드 수정
- StoreService
 : StoreListRespDto 사용 관련 코드 수정

TIP-Santa committed:
#64 test 리뷰 반영
- StoreServiceTest : StoreListRespDto 사용 관련 코드 수정

TIP-Santa committed:
dev pull > merge

yaewon0411 commented:
```
    @Query("SELECT sc.store.id AS storeId, sc.store.name AS storeName, sc.category.id AS categoryId " +
            "FROM StoreCategory sc WHERE sc.category.id = :categoryId")
    Page<StoreCategoryProjection> findStoresByCategoryId(Long categoryId, Pageable pageable);
```
위 쿼리에서는 프로젝션 dto의 minAmount를 조회하고 있지 않은데

서비스 단에서 아래와 같이 minAmount를 받아오고 있습니다

```
        Page<GetStoreRespDto> storeDtos = storeProjections.map(projection ->
                new GetStoreRespDto(projection.getStoreId(), projection.getStoreName(), projection.getMinOrderAmount()));
```

제 생각에 여기서 nullPointerException이 날 거 같은데 컨트롤러 WebMvcTest를 해보셔야 할 거 같습니다

TIP-Santa committed:
#64 refactor 리뷰 반영
- StoreCategoryRepository
 : 쿼리수정

TIP-Santa committed:
#75 refactor 리뷰 반영
- Category
 : Builder 추가
- GetStoreRespDto
 : 가게 조회 GetStoreRespDto 추가
- StoreCategoryProjection
 : getCategoryId 추가

TIP-Santa committed:
#75 test
- StoreControllerTest
 : 컨트롤러 단위 테스트 추가

TIP-Santa committed:
#75 test
- StoreControllerTest
 : 사용되지 않는 코드 정리

TIP-Santa committed:
dev pull > merge
